### ISSUE TITLE
Zap waiting for Spider endlessly when it crashed.

### ISSUE THREAD

**Issue by groulot** [zaproxy/zaproxy#4428]
I have an example reposity that has ZAP waiting for the Spider to end at 96% done. The spider never ends because apparently it crashes. Hitting CTRL-C displays an exception. In this case it is a HSQLDB problem:
`org.hsqldb.HsqlException: Data cache size limit is reached: 10000`

The issue is that ZAP doesn't detect this and just waits for the spider forever instead of failling or displaying an error.

Reproduction steps:

```
git clone https://gitlab.com/Flockademic/Flockademic.git
cd Flockademic
git checkout 1e6dbaf6 
mkdir wrk
docker run -ti --rm -v $PWD:/zap/wrk owasp/zap2docker-live zap-baseline.py -d -t https://flockademic.com/ 
```
It will start the spider and it quickly will get stuck at 96% . Interrupt with CTRL-c to see the exception. Here it is :

```
11512 [ZAP-SpiderInitThread-0] INFO org.zaproxy.zap.extension.spider.SpiderThread  - Starting spidering scan on https://flockademic.com at Thu Feb 22 13:29:31 UTC 2018
11514 [ZAP-SpiderInitThread-0] INFO org.zaproxy.zap.spider.Spider  - Spider initializing...
11550 [ZAP-SpiderInitThread-0] INFO org.zaproxy.zap.spider.Spider  - Starting spider...
27227 [ZAP-SpiderThreadPool-0-thread-1] INFO hsqldb.db..ENGINE  - dataFileCache CACHE SIZE limit reached
27228 [ZAP-SpiderThreadPool-0-thread-1] FATAL hsqldb.db..ENGINE  - /home/zap/.ZAP_D/session/untitled1.data getFromFile failed 37817
org.hsqldb.HsqlException: Data cache size limit is reached: 10000
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.persist.Cache.put(Unknown Source)
        at org.hsqldb.persist.DataFileCache.getFromFile(Unknown Source)
        at org.hsqldb.persist.DataFileCache.get(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.get(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.findNode(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.getRight(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.child(Unknown Source)
        at org.hsqldb.index.IndexAVL.insert(Unknown Source)
        at org.hsqldb.persist.RowStoreAVL.indexRow(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.indexRow(Unknown Source)
        at org.hsqldb.TransactionManager2PL.addInsertAction(Unknown Source)
        at org.hsqldb.Session.addInsertAction(Unknown Source)
        at org.hsqldb.Table.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementDML.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementInsert.getResult(Unknown Source)
        at org.hsqldb.StatementDMQL.execute(Unknown Source)
        at org.hsqldb.Session.executeCompiledStatement(Unknown Source)
        at org.hsqldb.Session.execute(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:350)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:295)
        at org.parosproxy.paros.model.HistoryReference.<init>(HistoryReference.java:360)
        at org.zaproxy.zap.extension.spider.SpiderThread.notifySpiderTaskResult(SpiderThread.java:488)
        at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:809)
        at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:241)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
27229 [ZAP-SpiderThreadPool-0-thread-1] ERROR org.zaproxy.zap.extension.spider.SpiderThread  - java.sql.SQLException: Data cache size limit is reached: 10000
org.parosproxy.paros.db.DatabaseException: java.sql.SQLException: Data cache size limit is reached: 10000
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:297)
        at org.parosproxy.paros.model.HistoryReference.<init>(HistoryReference.java:360)
        at org.zaproxy.zap.extension.spider.SpiderThread.notifySpiderTaskResult(SpiderThread.java:488)
        at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:809)
        at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:241)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.sql.SQLException: Data cache size limit is reached: 10000
        at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
        at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:350)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:295)
        ... 7 more
Caused by: org.hsqldb.HsqlException: Data cache size limit is reached: 10000
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.persist.Cache.put(Unknown Source)
        at org.hsqldb.persist.DataFileCache.getFromFile(Unknown Source)
        at org.hsqldb.persist.DataFileCache.get(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.get(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.findNode(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.getRight(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.child(Unknown Source)
        at org.hsqldb.index.IndexAVL.insert(Unknown Source)
        at org.hsqldb.persist.RowStoreAVL.indexRow(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.indexRow(Unknown Source)
        at org.hsqldb.TransactionManager2PL.addInsertAction(Unknown Source)
        at org.hsqldb.Session.addInsertAction(Unknown Source)
        at org.hsqldb.Table.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementDML.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementInsert.getResult(Unknown Source)
        at org.hsqldb.StatementDMQL.execute(Unknown Source)
        at org.hsqldb.Session.executeCompiledStatement(Unknown Source)
        at org.hsqldb.Session.execute(Unknown Source)
        ... 11 more
27230 [ZAP-SpiderThreadPool-0-thread-1] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-SpiderThreadPool-0-thread-1"
java.lang.NullPointerException
        at org.zaproxy.zap.extension.spider.SpiderScan.notifySpiderTaskResult(SpiderScan.java:339)
        at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:809)
        at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:241)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

Increasing the cache size allow the run to complete, but the underlying issue is still there.

```
docker run -ti --rm -v $PWD:/zap/wrk/:rw owasp/zap2docker-live sh -c 'sed -i "s/SET FILES CACHE SIZE 10000/SET FILES CACHE SIZE 100000/" db/zapdb.script && zap-baseline.py -d -t https://flockademic.com/ '
```

**Comment by thc202** [comment]
Thanks, yes, this should be handled better.

**Comment by zestysoft** [comment]
I just ran into this exact same problem with the 2.8.0 release.  Did something revert?

**Comment by thc202** [comment]
Could you provide more details?

**Comment by zestysoft** [comment]
I'm trying to think what I can add that isn't already here --  I attempted to spider, it got to around 86% and got stuck.  Console output showed data cache size limit exceptions:
```
Caused by: org.hsqldb.HsqlException: Data cache size limit is reached: 10000
```
Followed by:
```
501498 [ZAP-SpiderThreadPool-0-thread-2] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-SpiderThreadPool-0-thread-2"
java.lang.NullPointerException
	at org.zaproxy.zap.extension.spider.SpiderScan.notifySpiderTaskResult(SpiderScan.java:359)
	at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:845)
	at org.zaproxy.zap.spider.SpiderTask.runImpl(SpiderTask.java:248)
	at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:190)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

I bumped up the data cache size to fix.

**Comment by kingthorin** [comment]
> I bumped up the data cache size to fix.

That's fine as a work around, we've suggested that to people in the past. It's pretty rarely encountered (from my memory anyway).

**Comment by zestysoft** [comment]
What's odd is I didn't run into that in my previous month's scan or from any of the other scans I did using the weekly releases (or I didn't notice it because it somehow timedout and continued?).  That's why I thought maybe something reverted.

**Comment by zestysoft** [comment]
The one difference I can think of besides the little changes that happen to our web app each month is that amazon was blocking port 80 access and zap was having trouble connecting to the relevant urls.  Is it possible this could affect the data cache somehow?

**Comment by thc202** [comment]
> I'm trying to think what I can add that isn't already here

The question might seem pointless but I verified the fix then and just now with 2.8.0 and the spider does not get stuck:
```
87445 [ZAP-SpiderThreadPool-3-thread-2] ERROR org.zaproxy.zap.extension.spider.SpiderThread  - java.sql.SQLException: Data cache size limit is reached: 100
org.parosproxy.paros.db.DatabaseException: java.sql.SQLException: Data cache size limit is reached: 100
	at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:443)
	at org.parosproxy.paros.model.HistoryReference.<init>(HistoryReference.java:368)
	at org.zaproxy.zap.extension.spider.SpiderThread.notifySpiderTaskResult(SpiderThread.java:509)
	at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:845)
	at org.zaproxy.zap.spider.SpiderTask.runImpl(SpiderTask.java:248)
	at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:190)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.sql.SQLException: Data cache size limit is reached: 100
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
	at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:515)
	at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:427)
	... 8 more
Caused by: org.hsqldb.HsqlException: Data cache size limit is reached: 100
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.persist.Cache.put(Unknown Source)
	at org.hsqldb.persist.DataFileCache.add(Unknown Source)
	at org.hsqldb.persist.RowStoreAVLDisk.add(Unknown Source)
	at org.hsqldb.persist.RowStoreAVLDisk.getNewCachedObject(Unknown Source)
	at org.hsqldb.Table.insertSingleRow(Unknown Source)
	at org.hsqldb.StatementDML.insertSingleRow(Unknown Source)
	at org.hsqldb.StatementInsert.getResult(Unknown Source)
	at org.hsqldb.StatementDMQL.execute(Unknown Source)
	at org.hsqldb.Session.executeCompiledStatement(Unknown Source)
	at org.hsqldb.Session.execute(Unknown Source)
	... 12 more
87511 [ZAP-SpiderThreadPool-3-thread-2] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-SpiderThreadPool-3-thread-2"
java.lang.NullPointerException
	at org.zaproxy.zap.extension.spider.SpiderScan.notifySpiderTaskResult(SpiderScan.java:359)
	at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:845)
	at org.zaproxy.zap.spider.SpiderTask.runImpl(SpiderTask.java:248)
	at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:190)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
88900 [ZAP-SpiderThreadPool-3-thread-4] INFO org.zaproxy.zap.spider.Spider  - Spidering process is complete. Shutting down...
88901 [ZAP-SpiderShutdownThread-3] INFO org.zaproxy.zap.extension.spider.SpiderThread  - Spider scanning complete: true
```

So, there must be something slightly different causing the spider to get stuck in your case. I'll check the code and try to reproduce that.

> Followed by:

Is that the very last exception in the log? Is there a previous `NullPointerException`?
